### PR TITLE
Add evaluators for performing Poisson regression and Smoothed Hinge L…

### PR DESCRIPTION
**Pull request by ashelkovnykov**
…oss Linear SVM training in GAME

This change is necessary to enable Poisson Regression and Smoothed Hinge Loss for GAME.

### CODE REVIEW

**Comment by XianXing** [comment]
Can you update README.md to reflect these changes as well? For example, just mention that we have functionalities like model training with smoothed hinge loss and model training without regularization.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala [hunk 1]
```diff
@@ -287,9 +291,9 @@ final class Driver(val params: Params, val sparkContext: SparkContext, val logge
 
     val evaluator =
       taskType match {
-        case LOGISTIC_REGRESSION =>
+        case LOGISTIC_REGRESSION | SMOOTHED_HINGE_LOSS_LINEAR_SVM =>
           new BinaryClassificationEvaluator(validatingLabelAndOffsets)
-        case LINEAR_REGRESSION =>
+        case LINEAR_REGRESSION | POISSON_REGRESSION =>
```
Comment (ankans) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: After a discussion last time, we had decided to use the log likelihood as the best evaluator for Poisson Regression, given that it should always increase on the training data and generally increase on the test data as well. RMSE is clearly a bad metric, since Poisson regression does not try to fit the actual value of the predictor but a complicated function involving exp of the predictor.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/LogisticLossEvaluator.scala [hunk 2]
```diff
@@ -19,48 +19,49 @@ import org.apache.spark.rdd.RDD
...
     scoreAndLabelAndWeights.map { case (score, (label, weight)) =>
-      if (label > MathConst.POSITIVE_RESPONSE_THRESHOLD) {
-        weight * Utils.log1pExp(-score)
-      } else {
-        weight * Utils.log1pExp(score)
+        if (label > MathConst.POSITIVE_RESPONSE_THRESHOLD) {
+          weight * Utils.log1pExp(-score)
+        } else {
+          weight * Utils.log1pExp(score)
+        }
```
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/LogisticLossEvaluator.scala]: Can we call ```PointwiseLogisticLossFunction.loss``` directly to get the loss function value, instead of re-defining and computing the loss function value here?

The same goes for ```PoissonLossEvaluator```, ```SmoothedHingeLossEvaluator``` and ```SquaredLossEvaluator```.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    val scoreAndLabelAndWeights = scores
+      .rightOuterJoin(labelAndOffsetAndWeights)
+      .mapValues { case (scoreOption, (label, offset, weight)) =>
+        (scoreOption.getOrElse(defaultScore) + offset, (label, weight))
+      }
+      .values
+
+    scoreAndLabelAndWeights.map { case (score, (label, weight)) =>
+        weight * (math.exp(score) - score * label)
```
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala]: See comment above.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/SmoothedHingeLossEvaluator.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    scoreAndLabelAndWeights.map { case (score, (label, weight)) =>
+        val z = if (label < MathConst.POSITIVE_RESPONSE_THRESHOLD) -score else score
+
+        if (z <= 0) {
+          weight * (0.5 - z)
+        } else if (z < 1) {
+          weight * 0.5 * (1.0 - z) * (1.0 - z)
+        } else {
+          0
```
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/SmoothedHingeLossEvaluator.scala]: See comment above, but we may need some special treatment for ```SmoothedHingeLossFunction```.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/SquaredLossEvaluator.scala [hunk 5]
```diff
@@ -17,43 +17,45 @@ package com.linkedin.photon.ml.evaluation
...
+        }
+      .values
 
     scoreAndLabelAndWeights.map { case (score, (label, weight)) =>
-      val diff = score - label
-      weight * diff * diff
-    }.reduce(_ + _)
+        val diff = score - label
+        weight * diff * diff
+      }
```
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/SquaredLossEvaluator.scala]: Ditto.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala (hunk 1 above)
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: Yes, I had asked @ankans but forgotten to make the change, sorry about that

**Comment by ashelkovnykov** [comment]
While making the suggested changes, I've also enabled using validation record weights during evaluation (previously they were being ignored).

File: photon-ml/src/main/scala/com/linkedin/photon/ml/function/SmoothedHingeLossFunction.scala [hunk 6]
```diff
@@ -15,22 +15,42 @@
...
-    val z = actualLabel * margin
+    val (loss, deriv) = SmoothedHingeLossFunction.lossAndDerivative(margin, datum.label)
+
+    // Eq. 5, page 2 (derivative multiplied by label in lossAndDerivative method)
+    breeze.linalg.axpy(datum.weight * deriv, datum.features, cumGradient)
+    datum.weight * loss
+  }
+}
+
+object SmoothedHingeLossFunction {
```
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/function/SmoothedHingeLossFunction.scala]: Are these changes already covered by existing unit tests?

Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/function/SmoothedHingeLossFunction.scala]: Yes, these changes just move the calculation out to an object so that it is available for the `SmoothedHingeLossEvaluator`, they shouldn't modify the behaviour. Even so, the behaviour is tested by the `ObjectiveFunctionTest` and `ObjectiveFunctionIntegTest` tests.

**Comment by XianXing** [comment]
LGTM. Will merge if Travis is happy with this PR too.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    val scoreAndLabelAndWeights = scores
+      .rightOuterJoin(labelAndOffsetAndWeights)
+      .mapValues { case (scoreOption, (label, offset, weight)) =>
+        (scoreOption.getOrElse(defaultScore) + offset, (label, weight))
+      }
+      .values
+
+    scoreAndLabelAndWeights.map { case (score, (label, weight)) =>
+        weight * PointwisePoissonLossFunction.loss(score, label)._1
```
Comment (ankans) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala]: Are you using the same unit test for PoissonGradient evaluation that Brendan was using? Also, it is mandatory that the users only run Poisson regression with normalized features. Should this be specified as a comment in this file as well?
